### PR TITLE
Updates library to be compatible with PHP8.x

### PR DESCRIPTION
**Pull request by vanqard**
- updated implementations of built-in interface methods (Iterator, ArrayAccess and Countable)
 - renamed Match() class/file and updated references to it
 - deleted unnecessary empty files and fully deprecated stubs (Bool (query), Null (transport))

### CODE REVIEW

**Comment by vanqard** [comment]
These changes can't be brought in until the app is on PHP8 incidentally.

But to do so should just be a case of adding this to the `repositories` list in composer.json

```
        {
            "type": "github",
            "url": "git@github.com:roxhill/Elastica.git"
        }
```

**Review by mmieluch — CHANGES_REQUESTED** [review]
I have two gripes here.

1. I don't understand why we're updating Elastica to version 8, if Core is working in 7.4.
2. If we're truly embracing PHP 8 in this PR, I'd like to see strong types wherever possible. That means that some of the `mixed` types could be replaced with `int|string`, when typing offset accessor arguments for example.

Apart from that, the PR looks :ok_hand:

File: composer.json [hunk 1]
```diff
@@ -12,7 +12,7 @@
         }
     ],
     "require": {
-        "php": ">=5.6.0",
+        "php": "^8.0",
```
Comment (piotr0beschel) [inline composer.json:15]:
```suggestion
        "php": "^7.4|^8.0",
```

File: lib/Elastica/Bulk/ResponseSet.php [hunk 2]
```diff
@@ -106,38 +106,38 @@ public function current()
...
+    public function next(): void
     {
         ++$this->_position;
     }
 
     /**
      * @return int
      */
-    public function key()
+    public function key(): mixed
```
Comment (piotr0beschel) [inline lib/Elastica/Bulk/ResponseSet.php:117]: Return type does not match the docblock

File: lib/Elastica/Multi/ResultSet.php [hunk 3]
```diff
@@ -91,38 +91,38 @@ public function current()
...
+    public function next(): void
     {
         ++$this->_position;
     }
 
     /**
      * @return int
      */
-    public function key()
+    public function key(): mixed
```
Comment (piotr0beschel) [inline lib/Elastica/Multi/ResultSet.php:102]: Return type does not match the docblock

File: lib/Elastica/Multi/ResultSet.php [hunk 4]
```diff
@@ -132,7 +132,7 @@ public function count()
      *
      * @return bool true on success or false on failure.
      */
-    public function offsetExists($offset)
+    public function offsetExists(mixed $offset): bool
```
Comment (piotr0beschel) [inline lib/Elastica/Multi/ResultSet.php:135]: Why is this `mixed`?

File: lib/Elastica/Multi/ResultSet.php [hunk 5]
```diff
@@ -142,7 +142,7 @@ public function offsetExists($offset)
      *
      * @return mixed Can return all value types.
      */
-    public function offsetGet($offset)
+    public function offsetGet(mixed $offset): mixed
```
Comment (piotr0beschel) [inline lib/Elastica/Multi/ResultSet.php:145]: Why `mixed` and not `?int`

File: lib/Elastica/Bulk/ResponseSet.php (hunk 2 above)
Comment (mmieluch) [reply lib/Elastica/Bulk/ResponseSet.php:117]: Actually one might suggest removing docblocks without additional value altogether. I hopped on board of Jack's train of thought - it's only content duplication at this point, doesn't add any value.

File: composer.json (hunk 1 above)
Comment (mmieluch) [reply composer.json:15]: This will only apply if we keep the `mixed` types, and not embrace union types.

**Review by piotr0beschel — COMMENTED** [review]
Unit tests are failing

File: composer.json (hunk 1 above)
Comment (piotr0beschel) [reply composer.json:15]: I want it to support 7.4 as I want the core to run with update library before the upgrade to PHP 8.X

File: lib/Elastica/Bulk/ResponseSet.php (hunk 2 above)
Comment (piotr0beschel) [reply lib/Elastica/Bulk/ResponseSet.php:117]: I understand that docblocks are not adding anutything, but please left them be - to keep it as close to original lib as much as possible.

**Comment by vanqard** [comment]
Apologies @piotr0beschel  @mmieluch  - This is the earlier version that was php8 only. Let me get the one that's simultaneously 7 and 8 compatible up.

**Comment by vanqard** [comment]
New PR: https://github.com/roxhill/Elastica/pull/2

(This was the PHP8 only version)

**Pull request closed without merging**